_source/_posts/2021-10-16-full-stack-java.md

```diff
@@ -0,0 +1,691 @@
+    return `https://www.jhipster.tech/${claim}`;
+  }
+
+  const rolesClaim = prepareCustomClaimKey('roles');
+
+  if (context.idToken) {
+    context.idToken[rolesClaim] = roles;
+  }
+
+  if (context.accessToken) {
+    context.accessToken[rolesClaim] = roles;
+  }
+
+  callback(null, user, context);
+}
+```
+Click **Save changes** to continue.
+
+### Run Your JHipster App with Auth0
+
+Stop your JHipster app using **Ctrl+C**, set your Auth0 properties in `.auth0.env`, and start your app again.
+
+```shell
+source .auth0.env
+./mvnw
+```
+
+**NOTE:** If you're on Windows, you may need to install the [Windows Subsystem for Linux](https://docs.microsoft.com/en-us/windows/wsl/about) for this command to work.
+
+_Voilà_ - your full stack app is now using Auth0! Open your favorite browser to `http://localhost:8080`. 
+
+{% img blog/full-stack-java/jhipster-homepage.png alt:"JHipster default homepage" width:"800" %}{: .center-image }
+
+Click **sign in** and you'll be redirected to Auth0 to log in.
+
+{% img blog/full-stack-java/auth0-login.png alt:"Auth0 Login" width:"800" %}{: .center-image }
+
+After entering your credentials, you'll be redirected back to your app. 
+
+{% img blog/full-stack-java/jhipster-logged-in.png alt:"Authenticated!" width:"800" %}{: .center-image }
+
+## Test Your Full Stack Java App with Cypress
+
+JHipster has Auth0 support built-in, so you can specify your credentials for Cypress tests and automate your UI testing!
+
+To do this, specify the credentials for the Auth0 user you just created and run `npm run e2e`.
+
+```shell
+export CYPRESS_E2E_USERNAME=<new-username>
+export CYPRESS_E2E_PASSWORD=<new-password>
+npm run e2e
+```
+
+Everything should pass in around a minute.
+
+```shell
+       Spec                                              Tests  Passing  Failing  Pending  Skipped
+  ┌────────────────────────────────────────────────────────────────────────────────────────────────┐
+  │ ✔  administration/administration.spec.      00:22        5        5        -        -        - │
+  │    ts                                                                                          │
+  └────────────────────────────────────────────────────────────────────────────────────────────────┘
+    ✔  All specs passed!                        00:22        5        5        -        -        -
+
+Execution time: 59 s.
+```
+
+## Create Entities to allow CRUD on Photos
+
+I've talked a lot about how to secure your application, but we haven't done anything with photos! JHipster has a JDL (JHipster Domain Language) feature that allows you to model the data in your app, and generate entities from it. You can use its [JDL Studio](https://start.jhipster.tech/jdl-studio/) to do this online and save it locally once you've finished.
+
+My data model for this app has an `Album`, `Photo`, and `Tag` entities and sets up relationships between them. Below is a screenshot of what it looks like in JDL Studio.
```

Comment:
```suggestion
My data model for this app has `Album`, `Photo`, and `Tag` entities and sets up relationships between them. Below is a screenshot of what it looks like in JDL Studio.
```
removed "an" because entities is plural